Add debug section to Contributing.md (#1149)
github/gh-gei: CONTRIBUTING.md

All work done by the maintainers of this repo is tracked in this repo using Issues. We have a hierarchical backlog with Epics at the top, broken down into Batches then broken down to Tasks (epic/batch/task is indicated via labels on the issues). You can see an example Epic and navigate down from there [here](https://github.com/github/gh-gei/issues/101).

## Debugging code 

There are two ways to debug code within VS code. 

### Run GH-GHEI command locally
Run the following commands in your terminal depending on the provider you are looking to run the commands for. 

Azure DevOps: 

> `dotnet run --project src/ado2gh/ado2gh.csproj -- <<INSERT COMMAND HERE>>`

Reviewer: In most CLI documentation (i.e. man pages), straight brackets explicitly document optional parameters.

For example, here is `man ls`:

```
SYNOPSIS
       ls [OPTION]... [FILE]...
```

And here's `man top`:

```
SYNOPSIS
       top [options]
```

It would be more conformist to also use straight brackets here:

```suggestion
`dotnet run --project src/ado2gh/ado2gh.csproj -- [command]`
```

If you dig, let's also use this with the BBS and GitHub commands, too :+1:

Bitbucket Server: 
`dotnet run --project src/bbs2gh/bbs2gh.csproj -- <<INSERT COMMAND HERE>>`

GitHub:
`dotnet run --project src/gei/gei.csproj -- <<INSERT COMMAND HERE>>`

### Use debugger

If you use the built in debugger you are able to set breakpoints and inspect the code within VS Code. 

1. Navigate to `.vs_code/launch.json`.
2. Find the command you are looking to run, for example: `Launch ado2gh`.

> 3. Update the args property to have the arguments you're looking to test.
>     - "args": ["migrate-repo", "--ado-org", "<<INSERT_ORG>>", "--ado-team-project", "<<INSERT_PROJECT>>" ...]

Reviewer: How about:

```suggestion
3. Update the args property to have the arguments you're looking to test (using real org and project names):
    - "args": ["migrate-repo", "--ado-org", "example-org", "--ado-team-project", "example-project" ...]
```

All work done by the maintainers of this repo is tracked in this repo using Issues. We have a hierarchical backlog with Epics at the top, broken down into Batches then broken down to Tasks (epic/batch/task is indicated via labels on the issues). You can see an example Epic and navigate down from there [here](https://github.com/github/gh-gei/issues/101).

## Debugging code 

There are two ways to debug code within VS code. 

### Run GH-GHEI command locally
Run the following commands in your terminal depending on the provider you are looking to run the commands for. 

> Azure DevOps: 
> `dotnet run --project src/ado2gh/ado2gh.csproj -- [command]`

Author: ```suggestion
Azure DevOps: 
- Generic: `dotnet run --project src/ado2gh/ado2gh.csproj -- [command]`
- Example: `dotnet run --project src/ado2gh/ado2gh.csproj -- migrate-repo --ado-org "test-org" --ado-team-project "test-project" --ado-repo "test-ado-repo" --github-org "test-github-org" --github-repo "test-github-repo" `
```
@synthead I liked your idea about adding a real example here! How do you feel about this? Happy to reformat

Reviewer: Ooh yeah, I like it!  I would also remove the quotes in the example, since there aren't any special characters in your example.  Or alternatively, you could use a simple arg, like `--help` :slightly_smiling_face: Feel free to format it however it makes sense to you :+1:

Author: Updated!

Bitbucket Server: 
`dotnet run --project src/bbs2gh/bbs2gh.csproj -- [command]`

GitHub:
`dotnet run --project src/gei/gei.csproj -- [command]`

> 

Reviewer 2: A few days ago Arin helped get set up in a C# repl to test file uploads with our [AwsApi](https://github.com/github/gh-gei/blob/main/src/Octoshift/Services/AwsApi.cs). Not sure if this is the right place for it, but it might be useful to others debugging so thought I'd suggest.
````suggestion

### Run code in a C# REPL
In your terminal:
`dotnet tool install --global CSharpRepl --version 0.4.0`
(v0.4.0 is the latest version compatible with .NET 6.0)

Run it:
`csharprepl -r src/bbs2gh/bin/Debug/net6.0/Octoshift.dll`

Then load up assemblies:
```csharp
#r "Octoshift.dll"

// You might need others, for example the AWS SDK:
#r "AWSSDK.Core.dll"
#r "AWSSDK.S3.dll"

using OctoshiftCLI.Services;
```

````

Reviewer 3: This is going to be a great addition, just a few notes:
1 . Add some introduction, e.g. 

You can use a C# REPL to execute any C# code. There are many C# REPLs available, [CSharpRepl](https://fuqua.io/CSharpRepl/) is one of them and can be installed globally with the following:

`dotnet tool install --global CSharpRepl`

2. Not sure if the version is required.
3. I would add a comment before `using OctoshfitCLI.Services;`, something like this:
```csharp
// Add necessary usings
using OctoshiftCLI.Services

// Instantiate your classes
var aws = new AwsApi("access-key-id", "secret-access-key");
```

Author: Love this addition! Could you give a little more context of the situation it wold be used in?

Reviewer 3: @begonaguereca, it's a general purpose tool just like `irb` that gives you an interactive console to do whatever you like, for example instantiating `AwsApi` to test file i/o, etc.

Author: Thank you! Ok I just merged both of your suggestions together <3

### Use debugger

If you use the built in debugger you are able to set breakpoints and inspect the code within VS Code. 

1. Navigate to `.vs_code/launch.json`.
2. Find the command you are looking to run, for example: `Launch ado2gh`.
3. Update the args property to have the arguments you're looking to test.
    - "args": ["migrate-repo", "--ado-org", "<<INSERT_ORG>>", "--ado-team-project", "<<INSERT_PROJECT>>" ...]
4. Set a breakpoint within your code as needed.
5. Navigate to the `Run and debug` side panel option.
6. Navigate to the drop down menu and select the command you would like to run, for example: `Launch ado2gh`.
7. Press the play button

### Useful links 

1. [Run a .NET app](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-run)

> 2. [Debugging using vscode](https://code.visualstudio.com/docs/editor/debugging)

Author: @ArinGhazarian  thank you for pointing out helpful links!
